ssl/quic/quic_port.c

```diff
@@ -1336,6 +1357,183 @@ static int port_validate_token(QUIC_PKT_HDR *hdr, QUIC_PORT *port,
     return ret;
 }
 
+/*
+ * removes the mask from packet number field in QUIC packet header.
+ * the address to packet header is kept in ptrs structure.
+ */
+static int port_chk_remove_hdr_protection(QUIC_PORT_SECRETS *qps,
+                                          QUIC_PKT_HDR_PTRS *ptrs)
+{
+    int l = 0;
+    int ok;
+    unsigned char pn_len, i;
+    unsigned char dst[16] = { 0 };
+    unsigned char mask[5] = { 0 };
+    unsigned char *first_byte;
+    unsigned char *pn_bytes;
+
+    l = 0;
+    ok = EVP_CipherUpdate(qps->qps_hpr_ctx, dst, &l, ptrs->raw_sample, 16);
+    if (ok == 0)
+        return 0;
+    /*
+     * We are going to drop header protection here. We decode packet number
+     * using the mask we got from raw_sample deciphered by header protector
+     * key.
+     */
+    for (i = 0; i < 5; ++i)
+         mask[i] = dst[i];
+    first_byte = ptrs->raw_start;
+    pn_bytes = ptrs->raw_pn;
+
+    *first_byte ^= mask[0] & ((*first_byte & 0x80) != 0 ? 0xf : 0x1f);
+    pn_len = (*first_byte & 0x3) + 1;
+    for (i = 0; i < pn_len; ++i)
+        pn_bytes[i] ^= mask[i + 1];
+
+    return 1;
+}
+
+static int port_chk_derive_nonce(QUIC_PORT_SECRETS *qps, QUIC_PN pn,
+                                 unsigned char *nonce, int nonce_sz)
+{
+    int nonce_len;
+    size_t i;
+
+    nonce_len = EVP_CIPHER_CTX_get_iv_length(qps->qps_pkt_ctx);
+    if (nonce_len > nonce_sz)
+        return 0;
+
+    memcpy(nonce, qps->qps_rx_cipher_iv, nonce_len);
+    for (i = 0; i < sizeof(QUIC_PN); ++i)
+        nonce[nonce_len - i - 1] ^= (unsigned char)(pn >> (i * 8));
+
+    return 1;
+}
+
+/*
+ * Function verifies integrity of INITIAL QUIC packet.
+ */
+static int port_chk_initial_pkt(QUIC_PORT *port, QUIC_URXE *e,
+                                QUIC_PKT_HDR *hdr, QUIC_PORT_SECRETS **pqps)
+{
+    PACKET pkt;
+    QUIC_PKT_HDR  hdr_local = { 0 };
+    QUIC_PKT_HDR_PTRS ptrs = { 0 };
+    unsigned char *work_buf = NULL;
+    size_t work_buf_sz = 0;
+    unsigned char scratch_dst[65535];
```

Comment:
does this buffer need to be 65K?  I understand thats the maximum size of a UDP datagram, but IIRC QUIC packets can't be fragmented, implying that the largest URXE you receive here should be no more than the mtu of the link you're sending over, which I believe is available in port->demux->mtu at this time.  I get that that is a variable value and so would require a heap allocation here, so perhaps its not worthwhile, but the size of that value stood out to me, so I thought I would ask